src/util/domains.py

```diff
@@ -45,6 +43,95 @@
 except NameError:
     basestring = unicode = str
 
+# import from domains/expression
+try:
+    import odoo.domains as _dom
+
+    FALSE_LEAF = _dom._FALSE_LEAF
+    TRUE_LEAF = _dom._TRUE_LEAF
+    NOT_OPERATOR = _dom.DomainNot.OPERATOR
+    AND_OPERATOR = _dom.DomainAnd.OPERATOR
+    OR_OPERATOR = _dom.DomainOr.OPERATOR
+    DOMAIN_OPERATORS = {NOT_OPERATOR, AND_OPERATOR, OR_OPERATOR}
+
+    # normalization functions are redefined here because they will be deprecated
+    # the Domain factory normalized but also distributes operators during creation of domains
+
+    def normalize_domain(domain):
+        """Return a normalized version of the domain.
+
+        The normalized version have all implicit '&' operators explicit.
+        One property of normalized domain expressions is that they
+        can be easily combined together as if they were single domain components.
+        """
+        assert isinstance(
+            domain, (list, tuple)
+        ), "Domains to normalize must have a 'domain' form: a list or tuple of domain components"
+        if not domain:
+            return [TRUE_LEAF]
+        result = []
+        expected = 1
+        op_arity = {NOT_OPERATOR: 1, AND_OPERATOR: 2, OR_OPERATOR: 2}
+        for token in domain:
+            if expected == 0:  # more than expected, like in [A, B]
+                result[0:0] = [AND_OPERATOR]  # put an extra '&' in front
+                expected = 1
+            if isinstance(token, (list, tuple)):  # domain term
+                expected -= 1
+                if len(token) == 3 and token[1] in ("any", "not any"):
+                    new_token = (token[0], token[1], normalize_domain(token[2]))
+                    result.append(new_token)
+                else:
+                    result.append(normalize_leaf(token))
+            else:
+                expected += op_arity.get(token, 0) - 1
+                result.append(token)
+        if expected:
+            raise ValueError("Domain {!r} is syntactically not correct.".format(domain))
+        return result
+
+    def normalize_leaf(leaf):
+        if not is_leaf(leaf):
```

Comment:
If we allow in the linked PR in community to register optimizations that work accross all models we could do here:
`list(domains.Domain("a","<>",1)._optimize(None))`
which I can only seem to get to work with
```
list(domains.Domain("valid_field","<>",1)._optimize(self.env['base']))
```

Reply:
I don't think you should optimize domains during the migration, it does more than simple rewrites.
That operator is replaced automatically for a long time and continues to work for now.

Anyways, same remark as above.

Reply:
`self.env['base']` is a _particular_ model.
You need a different sentinel value for "any possible model".

Reply:
> You need a different sentinel value for "any possible model".

Indeed, to me it should be `None`. I was just hacking my way to get a feeling how this now works :)